[30577] Make work packages module to BCF module when activated

HDinger opened the pull request:
When activated the "Work Package" module shall be renamed to "BCF". In addition to the existing WP functionality, the `import`/ `export` button of the BCF module shall be added to the toolbar.

### ToDo
- [x] Rename WP Module when BCF is activated
- [x] Show import button in the toolbar, next to the create button (and make it work).
- [x] Show export button in the toolbar (and make it work)
- [x] Add a default BCF query, when the module is activated, which has the card view as default
- [x] Take care of correct redirections
  * Cancel of the import leads back to the list
  * A successful import leads to the "summary" page
  * The button on the summary page leads to the list 
- [x] Change the icon

https://community.openproject.com/projects/openproject/work_packages/30577/activity
https://community.openproject.com/projects/openproject/work_packages/30618/activity

HDinger committed:
Add BCF icon for menu entry

HDinger committed:
Rename "Work Packages" module to BCF when activated

HDinger committed:
Add import button for BCFs in the toolbar

HDinger committed:
Add BCF Export button the WP view

HDinger committed:
Add BCF default query and handle redirect after import

wielinde commented:
Hey @HDinger, thank you very much!! That PR will improve the UX of the BIM edition a lot.

Before starting the code review I opted for a quick "click through". Let me share some findings with you:

- During an import, when asked how to proceed with unknown entities such as Types, Statuses, Users and Priorities the Cancel-Buttons redirect to old controller index. It should redirect to the work packages module. Best would be to redirect back to the query where the user came from. I am not sure if that is easy to achieve. If not then simply go back to the work packages module's default view.
- After importing your PR directly redirects to the work package module. That happens one step too early. Before there was a view that showed the results of that import. And in case there were error messages associated with WP/BCF-issue, would get shown there. That view has a pretty hard to reach "Back to list" button which probably should be at the top of the page too and link back to work package module.
- Now seeing the buttons in the WP module's toolbar I think it is save to simply drop the "BCF-XML" from the export button label. I know I said something else before.

HDinger committed:
(1) Take care the "cancel" of import redirects to the WP list,
(2) A successful import redirects to the import summary page and not directly back to the list,
(3) Shorten button labels

HDinger commented:
@wielinde I included your feedback. For pragmatic reasons I have not implemented the redirect to the original query. Instead, it always returns to the default BCF query. I think the perfect redirect can still be done in a later step.

wielinde commented on app/helpers/application_helper.rb:
```diff
@@ -308,6 +308,8 @@ def body_css_classes
     css << "ee-banners-#{EnterpriseToken.show_banners? ? 'visible' : 'hidden'}"
 
+    css << "bcf-#{@project&.module_enabled?(:bcf) ? 'activated' : 'deactivated'}"
+
```
Any chance to move that logic to the BCF module? For example by patching the method?

wielinde commented on app/helpers/work_packages_filter_helper.rb:
```diff
@@ -214,6 +214,20 @@ def project_version_property_path(version, property_name, property_id, options =
     project_work_packages_with_query_path(version.project, query, options)
   end
 
+  def project_work_packages_bcf_issues_path(project)
```
Same as above. Any chance we can move that out of the core to maintain the separation of concerns?

wielinde commented on lib/redmine/plugin.rb:
```diff
@@ -273,6 +273,16 @@ def hide_menu_item(menu_name, item, hide_if: -> (*) { true })
+        menu_item.caption = options[:caption]
+        menu_item.icon = options[:icon]
+        menu_item.badge = options[:badge]
+
```
obsolete empty line

wielinde commented on modules/bcf/lib/open_project/bcf/engine.rb:
```diff
@@ -51,13 +51,12 @@ class Engine < ::Rails::Engine
-           param: :project_id,
-           icon: 'icon2 icon-backlogs',
-           badge: 'bcf.experimental_badge'
+      rename_menu_item :project_menu,
```
nice! Maybe we could even extend it to take a different URL to call? It would be great if this menu item and the view "All open" opens the grid view.

wielinde commented on lib/redmine/menu_manager/menu_item.rb:
```diff
@@ -88,6 +112,10 @@ def html_options(options = {})
     raise ArgumentError, 'Condition needs to be callable' unless new_condition.respond_to?(:call)
     old_condition = @condition
-    @condition = -> (project) { old_condition.call(project) && new_condition.call(project) }
+    if old_condition.respond_to?(:call)
```
I am curious why this change became necessary here. How can it be that this worked before?

HDinger replied:
It did not work before, when there was initially no condition given for the element. Since the method is not used very often, and those elements who used it have a condition, we didn't notice it.

wielinde reviewed (changes requested):
Hey @HDinger thank you for that great PR. I especially like how you decide on what is core and what is module. And sure, that separation is tricky in the frontend. Well done.

My comments are pretty small. So just have a quick look at them and then feel free to merge.

HDinger committed:
Override "all open" query when BCF is active and redirect to new "all open" when clicking on module name

HDinger committed:
Add BCF menu path to permission to see WP (otherwise the menu node won't pass the permission check and won't be visible)
